Part of #13551 - Create packages under release/debug directories as appropriate

leikahing opened the pull request:
r? @aneeshusa 

This change implements fixes for #13551 for Linux and MacOS targets.

`/python/servo/package_commands.py` was modified so that:
- On MacOS, it creates all intermediate packaging directories like `dmg`, `brew`, and `brew-tmp` under `target/(release|debug)`, rather than in `target` directly.
- On MacOS, all packaging artifacts (`.dmg`, brew `.tar.gz`) are packaged under `target/(release|debug)`, rather than in `target` directly.
- On Linux, the resulting `tar.gz` Servo package is placed under `target/(release|debug)`, rather than in `target`.
- Also did some extra cleanup around path parsing in the MacOS packaging code, to use `os.path` methods rather than straight `'/'` parsing with `split` and `join` where it was applicable.

`/etc/ci/upload_nightly.sh` was modified to:
- Look for artifacts in `target/release` for `mac`, `macbrew`, and `linux` platforms, rather than just `target/`.

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #13551
- [x] These changes do not require tests because it is based on CI and packaging tools. They were manually tested for correctness.

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

…lease/debug as specified or detected. Modify macos packaging to create all packages under release/debug directory. Updated etc/ci/upload_nightly.sh to support uploading from either release/debug directory, depending on what was built

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13845)

<!-- Reviewable:end -->

aneeshusa commented:
I've only skimmed the description so far, but thanks for the PR, and extra thanks for going above and beyond! Cleaning up the path manipulations in the OS X builders has been on my TODO list for a long time :)

KiChjang commented:
r? @aneeshusa

Highfive is slacking off it seems.

birryree commented:
Oh huh, thanks @KiChjang - I thought my original message assigned successfully.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -153,14 +153,14 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
         elif is_macosx():
 
-            dir_to_build = '/'.join(binary_path.split('/')[:-1])
+            dir_to_build = path.dirname(binary_path)
```
Please split out all of the hacking string processing -> `os.path` function changes into a separate commit.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -153,14 +153,14 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
-            dir_to_dmg = '/'.join(binary_path.split('/')[:-2]) + '/dmg'
-            dir_to_app = dir_to_dmg + '/Servo.app'
-            dir_to_resources = dir_to_app + '/Contents/Resources/'
+            dir_to_dmg = path.join(path.dirname(binary_path), 'dmg')
```
First arg here is `dir_to_build`.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -153,14 +153,14 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
-            dir_to_resources = dir_to_app + '/Contents/Resources/'
+            dir_to_dmg = path.join(path.dirname(binary_path), 'dmg')
+            dir_to_app = path.join(dir_to_dmg, 'Servo.app')
+            dir_to_resources = path.join(dir_to_app, 'Contents/Resources/')
```
`path.join(dir_to_app, 'Contents', 'Resources')`

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -211,10 +211,9 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
             time = now.replace(microsecond=0).isoformat()
             time = time.replace(':', '-')
-            dmg_path += time + "-servo-tech-demo.dmg"
+            dmg_path = path.join(dir_to_build, time + "-servo-tech-demo.dmg")
```
If it fits the line limit, I'd prefer string formatting for the second arg:
`'{}-servo-tech-demo.dmg'.format(time)`

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -318,7 +317,7 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
             time = datetime.utcnow().replace(microsecond=0).isoformat()
             time = time.replace(':', "-")
-            tar_path = path.join(self.get_target_dir(), time + '-servo-tech-demo.tar.gz')
+            tar_path = path.join(os.path.dirname(binary_path), time + '-servo-tech-demo.tar.gz')
```
Same comment about using string formatting here if possible.

aneeshusa reviewed (changes requested):
Looks good overall. I'll take a closer look at the path manipulation changes once they're split up. It looks like you're going to earn a lot of bonus points on this PR!

aneeshusa commented on etc/ci/upload_nightly.sh:
```diff
@@ -25,9 +25,8 @@ upload() {
     s3cmd cp "${package_upload_path}" "${nightly_upload_dir}/servo-latest.${3}"
 }
 
-
```
Keep this line to avoid diff noise and fit with the double newline between functions style.

birryree replied:
This is resolved in the next commit (haven't done any squashing yet in order to separate out my fixes and improvements).

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -158,7 +158,7 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
 
             print("Creating Servo.app")
-            dir_to_dmg = '/'.join(binary_path.split('/')[:-2]) + '/dmg'
+            dir_to_dmg = '/'.join(dir_to_build) + '/dmg'
```
I think this should be `path.join(dir_to_build, 'dmg')`.

birryree replied:
Yes, I have a second commit incoming that does all the `path` tidying - just testing it all to make sure it works again as a blob (and because `servo` currently isn't building on my Mac for some reason).

aneeshusa reviewed (changes requested):
I really liked the path cleanups you had previously - sorry if I wasn't clear but I still want them in this PR, just as a separate commit to avoid this directory change from getting lost.

Also, for the brew package it looks like the upload script has the new location but the packaging code hasn't been changed (although I seem to remember it being there before).

leikahing committed:
Cleanup path manipulations under macOS/Linux

Path manipulation has been changed from using '/' to using
os.path functions for packaging code for macOS and Linux targets.

birryree commented:
@aneeshusa - I committed all my path changes again - sorry for the confusion. I had to redo my commits to make them separate and since I have to build on two different machines, I committed to my fork before re-adding all my path modifications back in (now with _more_ cleanups).

I re-tested everything on Linux/macos with all these changes - everything looks good on my end. Please re-review when you can. 😄

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14058) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Review ping @aneeshusa.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -152,13 +152,13 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
-            dir_to_build = '/'.join(binary_path.split('/')[:-1])
-            dir_to_root = '/'.join(binary_path.split('/')[:-3])
+            dir_to_build = path.dirname(binary_path)
+            dir_to_root = os.sep.join(binary_path.split(os.sep)[:-3])
```
You should be able to use `self.get_top_dir()` to get `dir_to_root`.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -168,19 +168,23 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
 
             print("Finding dylibs and relinking")
-            copy_dependencies(dir_to_app + '/Contents/MacOS/servo', dir_to_app + '/Contents/MacOS/')
+            copy_dependencies(path.join(dir_to_app, 'Contents', 'MacOS', 'servo'), content_dir)
```
You can reuse `content_dir` here inside the `path.join`.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -193,24 +197,25 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
             print("Writing run-servo")
-            bhtml_path = path.join('${0%/*}/../Resources', browserhtml_path.split('/')[-1], 'out', 'index.html')
-            runservo = os.open(dir_to_app + '/Contents/MacOS/run-servo', os.O_WRONLY | os.O_CREAT, int("0755", 8))
+            bhtml_path = path.join('${0%/*}/../Resources', browserhtml_path.split(os.sep)[-1], 'out', 'index.html')
```
Also, see if you can get rid of the explicit `/`s in the first `path.join` in favor of multiple args.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -193,24 +197,25 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
-            bhtml_path = path.join('${0%/*}/../Resources', browserhtml_path.split('/')[-1], 'out', 'index.html')
-            runservo = os.open(dir_to_app + '/Contents/MacOS/run-servo', os.O_WRONLY | os.O_CREAT, int("0755", 8))
+            bhtml_path = path.join('${0%/*}/../Resources', browserhtml_path.split(os.sep)[-1], 'out', 'index.html')
+            runservo = os.open(path.join(dir_to_app, 'Contents', 'MacOS', 'run-servo'),
```
Reuse `contents_dir`.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -221,24 +226,24 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
             print("Creating brew package")
-            dir_to_brew = '/'.join(binary_path.split('/')[:-2]) + '/brew_tmp/'
-            dir_to_tar = '/'.join(dir_to_build.split('/')[:-1]) + '/brew/'
+            dir_to_brew = path.join(path.dirname(binary_path), 'brew_tmp')
```
Reuse `dir_to_build`.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -221,24 +226,24 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
-            shutil.copy2(dir_to_build + '/servo', dir_to_brew + '/bin/servo')
+            shutil.copytree(path.join(dir_to_root, 'resources'), path.join(dir_to_brew, 'resources'))
+            os.makedirs(path.join(dir_to_brew, 'bin'))
+            shutil.copy2(path.join(dir_to_build, 'servo'), path.join(dir_to_brew, 'bin', 'servo'))
```
Reuse `binary_path`.

aneeshusa reviewed (changes requested):
Linux changes looks good to me, left some style comments about the OS X changes.

Please squash into two commits, the first doing general cleanups (i.e. using `os.path` functions), and the second commit only making the actual change of package directory.

Also, a few nits about commit message style:
commit messages:
- No need to mention the issue # in the commit title
- Ensure titles are less than ~50 chars
- Leave extra explanation in the commit message body as full sentences
  I like http://chris.beams.io/posts/git-commit/ as a guide for this.

birryree commented:
@aneeshusa Thanks again for feedback - I've cleaned up the code, squashed my commits into the 'feature change' commit and an enhancement commit, and changed my commit messages to conform to servo standards.

birryree commented:
Review ping @aneeshusa

aneeshusa commented:
Sorry about the delay. You've got all the rights bits in terms of code and commit messages, but they need to be shuffled around a bit to match.

- Switch the two commits so the first commit does all the cleanups.
  This will make it easier to read the other commit, which changes the actual locations.
- There are three changes that you need to migrate from 7a56f84 to f03cf09,
  to completely separate cleanups from semantic changes. Each of these is a change of the path some variable refers to, and all of these are in the macOS packager:
  - `dmg_path`, which is on line 217
  - `dir_to_brew`, which is on line 229
  - `dir_to_tar`, which is on line 230

leikahing committed:
Package macos/Linux builds under release/debug dirs

This commit updates the 'mach package' command to generate output
under the target/release and target/debug directories when run on
macOS and Linux.

etc/ci/upload_nightly.sh has also been updated to upload packages from
the release/debug directories.

birryree commented:
@aneeshusa Sorry for how long it's taken to get this PR in.

I've reordered my commits so that the path manipulation cleanups are the first commit.

I imagine your statement here:

> There are three changes that you need to migrate from 7a56f84 to f03cf09

Was for the other direction, so the path manipulation commit now changes the `os.path` usage based on the original version of `package_commands.py`, and then the second commit to change the packaging structure improves upon that commit.

aneeshusa commented on python/servo/package_commands.py:
```diff
@@ -193,24 +197,25 @@ def package(self, release=False, dev=False, android=None, debug=False, debugger=
-            bhtml_path = path.join('${0%/*}/../Resources', browserhtml_path.split('/')[-1], 'out', 'index.html')
-            runservo = os.open(dir_to_app + '/Contents/MacOS/run-servo', os.O_WRONLY | os.O_CREAT, int("0755", 8))
+            bhtml_path = path.join('${0%/*}', '..', 'Resources', path.basename(browserhtml_path), 'out', 'index.html')
+            runservo = os.open(path.join(content_dir, 'run-servo'),
```
nit: use block indentation:
```
runservo = os.open(
    path.join(...),
    os.O_WRONLY...,
    int("0755", 8)
)
```

aneeshusa reviewed (changes requested):
Looks great! One last style nit.

aneeshusa commented:
Everything looks good to me, thanks again for the PR and cleaning up all that path code! @bors-servo r+

bors-servo commented:
:pushpin: Commit 6a7fb6c has been approved by `aneeshusa`

<!-- @bors-servo r=aneeshusa 6a7fb6cd2b668a2da6d3abb909742948745ca058 -->

bors-servo commented:
:hourglass: Testing commit 6a7fb6c with merge 4cb3404...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/4563), [arm64](http://build.servo.org/builders/arm64/builds/4571), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4653), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1330), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1349), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4628), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4651), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1340), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1354), [windows-dev](http://build.servo.org/builders/windows-dev/builds/2595)
